### PR TITLE
add topn closure executor

### PR DESCRIPTION
**Pull request by bobotu**
implement closure executor for `TopN` operator

### CODE REVIEW

File: tikv/closure_exec.go [hunk 1]
```diff
@@ -157,6 +160,42 @@ func (svr *Server) tryBuildAggClosureExecutor(e *closureExecutor, executors []*t
 	return e, nil
 }
 
+func (svr *Server) tryBuildTopNClosureExecutor(e *closureExecutor, executors []*tipb.Executor) (*closureExecutor, error) {
+	if executors[0].Tp != tipb.ExecType_TypeTableScan {
```
Comment (jackysp) [inline tikv/closure_exec.go]: Should we check the size of `executors` here?

File: tikv/executor.go [hunk 2]
```diff
@@ -575,12 +575,16 @@ func (e *selectionExec) Next(ctx context.Context) (value [][]byte, err error) {
 	}
 }
 
-type topNExec struct {
+type baseTopNExec struct {
```
[the remark this answers is not in the document]
Comment (bobotu) [reply tikv/executor.go]:
```go
type baseTopNExec struct {
	heap              *topNHeap
	relatedColOffsets []int
	orderByExprs      []expression.Expression
	row               []types.Datum
}
```

`topNExec` also use these fields, and use the same init logical for them.

[the remark this answers is not in the document]
Comment (jackysp) [reply tikv/executor.go]: Both `topNExec` and `topNCtx` use `baseTopNExec`. It seems `topNCtx` is really a heavy context, it looks like an executor. Could we simplify `topNCtx`?

[the remark this answers is not in the document]
Comment (bobotu) [reply tikv/executor.go]: Ok, I'll push a new commit later, do some refactor to address comments above.

**Comment by bobotu** [comment]
### About `topNCtx.chk` and `scanCtx.chk`

During table scan process we don't need to decode all cols into chunk, we can decode the columns used in `order by`. The new row format can fast decode subset of columns.

During eval of expression in `order by`, we still need a `Chunk` respect to table's schema, even though we will only access the decoded columns.

So, I add a new method to `Chunk`, `ProjectColumns`. This method returns a new Chunk, which columns is simply a reference of the origin Chunk.
`scanCtx.chk` is returned by `topNCtx.chk.ProjectColumns()`, so after `tableScanProcessCore` decoded value, `Expr.eval` can use `topNCtx.chk.GetRow(0)` to get decoded row directly.

I'll move `ProjectColumns` to tidb later.

**Comment by bobotu** [comment]
@jackysp PTAL

**Pull request merged**